src/BrainPass/BrainPass.sol

```diff
@@ -0,0 +1,284 @@
+    mapping(address => mapping(uint256 => UserPassItem))
+        public addressToNFTPass;
+    mapping(address => mapping(uint256 => bool)) public addressToPassId;
+
+    /// -----------------------------------------------------------------------
+    /// Constant
+    /// -----------------------------------------------------------------------
+    address public IqToken;
+
+    /// -----------------------------------------------------------------------
+    /// Variables
+    /// -----------------------------------------------------------------------
+    string public baseTokenURI;
+    Counters.Counter private tokenIds;
+
+    /// -----------------------------------------------------------------------
+    /// Constructor
+    /// -----------------------------------------------------------------------
+    constructor(
+        address IqAddr
+    ) ERC721("BRAINY EDITOR PASS", "BEP") Owned(msg.sender) {
+        IqToken =IqAddr;
+    }
+
+    function baseURI() internal view virtual returns (string memory) {
+        return baseTokenURI;
+    }
+
+    function setBaseURI(string memory tokenURI) internal {
+        baseTokenURI = tokenURI;
+    }
+
+    /// @notice Add a new Pass Type
+    /// @param passId and others are the details needed for a passType
+    function addPassType(
+        uint256 passId,
+        uint256 pricePerDay,
+        string memory tokenURI,
+        string memory name,
+        uint256 maxTokens,
+        uint256 discount
+    ) public onlyOwner {
+        require(bytes(tokenURI).length > 0, "Invalid token URI");
+        require(maxTokens > 0, "Invalid max tokens");
+
+        passTypes[passId] = PassType(
+            passId,
+            name,
+            pricePerDay,
+            tokenURI,
+            maxTokens,
+            discount,
+            0
+        );
+
+        emit NewPassAdded(passId, name, maxTokens, pricePerDay);
+    }
+
+    /// @notice Mint and NFT of a particular passtype
+    /// @param passId The id of the passtype to mint
+    function mintNFT(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) public {
+        require(
+            addressToPassId[msg.sender][passId] != true,
+            "Max NFTs per address reached"
+        );
+
+        PassType storage passType = passTypes[passId];
+
+        require(passType.maxTokens != 0, "Pass type not found");
+
+        require(
+            passType.lastTokenIdMinted.add(1) <= passType.maxTokens,
+            "Max supply reached"
+        );
+
+        uint256 price = calculatePrice(passId, startTimestamp, endTimestamp);
+
+        if (IERC20(IqToken).balanceOf(msg.sender) < price)
+            revert UserBalanceNotEnough();
+
+        bool success = IERC20(IqToken).transferFrom(msg.sender, owner, 1e18);
+        if (!success) revert MintingPaymentFailed();
+
+        uint256 tokenId = passType.lastTokenIdMinted + 1;
+        setBaseURI(passType.tokenURI);
```

Comment:
arent you setting baseUri for all the tokens each time someone buy a passtype different?